### ISSUE TITLE
The .commands file is getting corrupted

### ISSUE THREAD

**Issue by jonathanrissler** [tig/mcec#24]
**Describe the bug**

_We use only the "shutdown"-command in MCE Controller. After a period of time, MCE Controller won't load the commands correctly, so "shutdown" is no longer enabled and therefore the PC won't shut down._

**To Reproduce**

_We haven't found a way yet to reproduce the error, but think it may be related to power failure._

**Expected behavior**

_We expected MCE Controller to shut the PC down, as it always had done. We use the tool on 50 PCs, and have the problem only on 5 devices. The error started occuring after 2 months._

**Screenshots**

![MCE_Controller_Fehlermeldung](https://user-images.githubusercontent.com/97680198/149342610-925a7fa2-4fc9-42a5-8375-095abe5a14bf.png)

**Version Information**

_MCE Controller v2.2.7.67 - OS: Microsoft Windows NT 10.0.17763.0 on x64 - .NET: 4.0.30319.42000_

**Snippets from the log**

```
Paste Logs here.
[MCEControl_6206-PC-01.log](https://github.com/tig/mcec/files/7863260/MCEControl_6206-PC-01.log)

```

**Desktop (please complete the following information):**
_Microsoft Windows 10 Pro	
Version	10.0.17763 Build 17763_	

**Additional context**

**Comment by tig** [comment]
Hello.

The log screenshot you shared does not match the contents of the .log file you linked to. In the .log file I don't see this error happening. To help me debug this, can you get me a .log file that shows the operations before and after the shutdown where the .commands file was lost/corrupted?

**Comment by tig** [comment]
Also, is there something about your network connection which is preventing MCE Controller from checking for new versions?

```
2022-01-11 13:31:26,267 DEBUG - Checking for new release...
2022-01-11 13:31:26,269 DEBUG - (https://github.com/tig/mcec/releases) Fehler beim Senden der Anforderung.
2022-01-11 13:31:26,275 INFO  - You are are running a MORE recent version than can be found at tig.github.io/mcec (0.0)
```

I don't think this could be related, but I'd like to ensure I don't have a bug in the update code. Thanks.

**Comment by jonathanrissler** [comment]
Hello,

that's right, sorry. I made the screenshot in November, on a different device than the one I took the .log-file from. 
And also right, the-log-file doesn't show the error. It starts in January, I don't know why that is.

There are no more operations to see before the error occurs. Everytime it happens (all in all 15 times by now) it's right after the start, where it says in the log "Error parsing .commands XML. Das Stammelement ist nicht vorhanden. (The trunk element does not exist.)"

Regarding your other question: Our PCs do not have internet connection and are supposed not to have. We downloaded the most recent version.

Thanks for your quick reply!

**Comment by tig** [comment]
Can you please share a full .log file from one of the PCs where the error has occured?

**Comment by tig** [comment]
Also, in the screenshot it appears you edited the .commands file and saved it manually. Is that correct?

**Comment by tig** [comment]
One more request:

Next time this happens, please check to see if `C:\Users\kiosk\AppData\Roaming\Kindel Systems\MCE Controller\MCEControl.commands` exists. 

If it does, please send me a copy.

If it does not, please let me know.

**Comment by jonathanrissler** [comment]
Yes, I will send you a .log file of one of the error-PCs. Unfortunately, I can only access them on Tuesdays, since they are part of a running installation.

No, I didn't save the .commands file manually.

Yes, next time I will check if the .commands file exists, it really seems like it either doesn't exist and is newly created or it is corrupt and therefore "lost" the one defined command.

**Comment by tig** [comment]
I've been travelling so have not been able to check in on this. I'm now in a more stable location and have time today. If you can get met the logs and other info, I should be able to focus on helping fix this! Let me kow.

**Comment by jonathanrissler** [comment]
Hi, no problem!
[MCEControl.log](https://github.com
[MCEControl.log](https://github.com/tig/mcec/files/7942821/MCEControl.log)
/tig/mcec/files/7942820/MCEControl.log)

I'm sorry, but again, there is nohing to see in the logs - they seem to "restart" after the error occurs, and so now when I checked, they didn't pass back until the time of the error (last November). You can see it only on the screenshot I put in the first description.
If you need anything more, please let me know.
Thanks.

**Comment by jonathanrissler** [comment]
Hi!
So now I've managed to catch two PCs in the act of the error. I collected all logs and error messages. Hope this helps!

[MCE_error_1_6006](https://user-images.githubusercontent.com/97680198/151969217-55714f40-a6b5-4d00-b48a-0ed6b9814093.PNG)
![MCE_error_2_6006](https://user-images.githubusercontent.com/97680198/151969222-8f36f1b3-2888-4861-b15c
![MCE_error_1_6126](https://user-images.githubusercontent.com/97680198/151969282-0cd82fe0-e46d-4393-b11e-ce3f2dc7f23e.PNG)
![MCE_error_2_6126](https://user-images.githubusercontent.com/97680198/151969283-615fc9e1-8748-4dd5-a0a2-8cb82f85ffd9.PNG)
-53fd221e7fe6.PNG)

**Comment by jonathanrissler** [comment]
[MCEControl_6126.log](https://github.com/tig/mcec/files/7978461/MCEControl_6126.log)
[MCEControl_6006.log](https://github.com/tig/mcec/files/7978462/MCEControl_6006.log)

I copied the complete MCE-folder, if you also need the .commands and -settings files. I can't paste them here but can send you a link or mail. Let me know. Thank You!

**Comment by tig** [comment]
> [MCEControl_6126.log](https://github.com/tig/mcec/files/7978461/MCEControl_6126.log) [MCEControl_6006.log](https://github.com/tig/mcec/files/7978462/MCEControl_6006.log)
> 
> I copied the complete MCE-folder, if you also need the .commands and -settings files. I can't paste them here but can send you a link or mail. Let me know. Thank You!

These two log files appear to be from Sept 2021. Did you capture any log files from around Feb 01 when the screen shots above were taken?

**Comment by jonathanrissler** [comment]
You're right, I didn't check on this, because I captured them just on Tuesday, when I took the screenshots. I checked on the other log files from PCs withour the error, and they all date back to August/September 2021, although I captured them just this year. Seems like the log isn't written over the entire period?

**Comment by tig** [comment]
> 

This is really concerning. I can't understand how recent logs are not being written. 

Can you please exit and restart MCE controller on one of the machines that is working ok and then send me

1) the latest .log file
2) a screen shot of the directory the .log, .settings, and .commands files are in

**Comment by tig** [comment]
In addition to doing the above, please:

1) Update all your machines to the new release: v2.2.8 (https://github.com/tig/mcec/releases/tag/v2.2.8.63)
2) In Settings, change the logging threshold to "ALL"

<img width="227" alt="image" src="https://user-images.githubusercontent.com/585482/152697760-6f602623-68fe-40b7-a683-b5da9fd11ba7.png">

This new release has better logging and I tweaked some things that MIGHT fix this problem.

However, without full logs, I can't know for certain why the .commands file is getting written as empty.

**Comment by jonathanrissler** [comment]
OK, I can do that tomorrow.

**Comment by jonathanrissler** [comment]
[MCEControl.log](https://github.com/tig/mcec/files/8022006/MCEControl.log)

The log file from one of the working machines - it just dates back to January 17.

![directory](https://user-images.githubusercontent.com/97680198/152947947-7676f062-9520-4ffc-8191-a50db189847b.PNG)

And the screenshot from the directory. It starts with C:\Users\kiosk

**Comment by jonathanrissler** [comment]
I updated MCE Controller on some of the machines an tested several times. Using another strategy, I wrote a batch file that always copies the correct .commands file to the right directory when the PC boots. Both seems to work, but if it will for a long period of time, we can only observe.
I will keep you informed, thanks for the Update!

**Comment by tig** [comment]
> [MCEControl.log](https://github.com/tig/mcec/files/8022006/MCEControl.log)
> 
> The log file from one of the working machines - it just dates back to January 17.
> 
> ![directory](https://user-images.githubusercontent.com/97680198/152947947-7676f062-9520-4ffc-8191-a50db189847b.PNG)
> 
> And the screenshot from the directory. It starts with C:\Users\kiosk

Thanks. FWIW, the `MCEControl.log.n` files are previous logs. At 100k file size, the logger starts a new file, saving the previous as `MCEControl.log.n`.  

Please send me ALL files in that directory. Thanks.

**Comment by jonathanrissler** [comment]
Ok. As usual, I can give you these files next Tuesday.

**Comment by tig** [comment]
Closing as resolved. As of the exit-time save being removed (`src/MainWindow.cs:485-486`, currently a `//Invoker.Save(...)` left in place with a `BUGBUG` comment referencing this issue directly), MCEC no longer writes `mcec.commands` automatically on every app exit — that unattended write racing a power loss or forced shutdown was the most plausible cause of the corruption reported here. The file is now only written from an explicit user action (the Commands window's "Save" button) or a one-time version-upgrade path.

Caveat: the write itself (`SerializedCommands.SaveCommands`) is still not crash-atomic, so an interruption at exactly the wrong moment during a manual save could still in theory corrupt the file. If anyone can reproduce corruption on a current v3.0.x build, please open a fresh issue with logs from immediately before/after.

(Also caught in the same investigation: `docs/home-automation.md` still claims the file is "saved automatically whenever MCEC exits" — filing a doc fix for that separately.)
